compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/SingleAbstractMethodLowering.kt

```diff
@@ -0,0 +1,237 @@
+                // Implicit conversion to an interface:
+                //    // FILE: 1.java
+                //    class A { public static void f(Runnable r) { r.run() } }
+                //    // FILE: 2.kt
+                //    A.f { print("Hello, World!") }
+                return createInstance(expression.typeOperand, expression.argument.transform(this, null))
+            }
+
+            // Replace a SAM-conversion of a callable object with an instance of an appropriate implementation.
+            private fun createInstance(superType: IrType, invokable: IrExpression): IrExpression {
+                // Running in the class context, so empty scope means this is a field/anonymous initializer.
+                val scopeOwnerSymbol = currentScope?.scope?.scopeOwnerSymbol ?: irClass.symbol
+                return context.createIrBuilder(scopeOwnerSymbol).irBlock(invokable, null, superType) {
+                    // Not an exhaustive check, but enough to cover the most common cases.
+                    if (invokable is IrFunctionReference) {
+                        // Runnable(::function)
+                        createFunctionProxyInstance(superType, invokable)
+                    } else if (invokable is IrBlock && invokable.statements.last() is IrFunctionReference) {
+                        // Runnable { lambda }
+                        for (statement in invokable.statements.dropLast(1))
+                            +statement
+                        createFunctionProxyInstance(superType, invokable.statements.last() as IrFunctionReference)
+                    } else {
+                        // Fall back to materializing an invokable object.
+                        createObjectProxyInstance(superType, invokable)
+                    }
+                }
+            }
+
+            // Construct a class that implements the specified SAM interface and contains a bunch of fields
+            // in which the arguments to the constructor are stored:
+            //     class sam$n(private val arg0: T0, ...) { override fun method(...) = <whatever buildOverride returns> }
+            private fun implement(
+                superType: IrType,
+                parameters: List<IrType>,
+                buildOverride: IrSimpleFunction.(List<IrField>) -> IrBody
+            ): IrClass {
+                val superClass = superType.classifierOrFail.owner as IrClass
+                // The language documentation prohibits casting lambdas to classes, but if it was allowed,
+                // the `irDelegatingConstructorCall` in the constructor below would need to be modified.
+                assert(superClass.kind == ClassKind.INTERFACE) { "SAM conversion to an abstract class not allowed" }
+
+                val superClassName = superClass.fqNameSafe.pathSegments().joinToString("_") { it.toString() }
+                val subclass = buildClass {
+                    // TODO this is not the name some tests (e.g. kt11519) expect
```

Comment:
I suppose we could extract a bytecode text test out of that test, to prevent looking up a class by a non-stable name. Or maybe try and look up both names to make that test pass in JVM IR